Map iteration support for ForEach processor

danhermann opened the pull request:
Adds support to the ForEach processor for iterating over map fields in addition to array fields. When iterating over maps, an additional ingest metadata field `_ingest._key` is populated with the value of the map entry's key. Both the map entry's key and value can be changed. Map entries can also be removed completely by setting the key to a null or empty string value.

Resolves https://github.com/elastic/elasticsearch/issues/55215

danhermann committed:
adds map iteration to ForEach processor

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Ingest)

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 55215_foreach_over_maps

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 55215_foreach_over_maps

martijnvg commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/ForEachProcessorTests.java:
```diff
@@ -311,6 +312,56 @@ public void testRemovingFromTheSameField() {
+        assertThat(ingestDocument.getFieldValue("field", Map.class).entrySet().toArray(),
+            arrayContainingInAnyOrder(Map.entry("foo", 1), Map.entry("baz", 3)));
+    }
+
```
Maybe also add a nested for each tests for maps like the existing `testNestedForEach()` test?

martijnvg commented on modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ForEachProcessor.java:
```diff
@@ -67,20 +67,61 @@ boolean isIgnoreMissing() {
             }
+        } else if (o instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map<String, ?> map = (Map<String, ?>) o;
```
I think we can cast to `Map<?, ?>` instead? Then the SuppressWarnings can be removed. In order to get the key a cast to String would required, but that shouldn't generate warnings.

martijnvg reviewed (commented):
Looks good, left 2 comments.

danhermann committed:
remove unchecked cast

danhermann committed:
wip on nested maps

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 55215_foreach_over_maps

danhermann committed:
Merge branch '55215_foreach_over_maps' of https://github.com/danhermann/elasticsearch into 55215_foreach_over_maps

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 55215_foreach_over_maps

danhermann committed:
Merge branch '55215_foreach_over_maps' of https://github.com/danhermann/elasticsearch into 55215_foreach_over_maps

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 55215_foreach_over_maps

danhermann committed:
Merge branch '55215_foreach_over_maps' of https://github.com/danhermann/elasticsearch into 55215_foreach_over_maps

danhermann committed:
fix nested map iteration and add tests

danhermann committed:
simplify tests

danhermann commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 55215_foreach_over_maps

danhermann commented:
@martijnvg, it took a long time 😞, but I fixed iteration over nested maps (two-line change [here](https://github.com/elastic/elasticsearch/pull/64062/commits/207982e8b374519ff9691c06a3e71f97f8f62a55#diff-d8bf82d8f0d8bfd971c61a0fc6a14671e1466e78951cbfc103b5de3fa882857c)) and added two test cases to verify that.

martijnvg reviewed (approved):
LGTM

danhermann commented:
Thanks, @martijnvg!
